File: Unknownbank/app/src/main/java/com/kasatria/kasatriaunknownbank/CreditCardFilterScreen.kt
package com.kasatria.kasatriaunknownbank

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.statusBarsPadding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material.icons.filled.KeyboardArrowUp
import androidx.compose.material.icons.filled.Search
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

import com.kasatria.kasatriaunknownbank.ui.theme.LinkBlue
import com.kasatria.kasatriaunknownbank.ui.theme.PrimaryBlue
import com.kasatria.kasatriaunknownbank.ui.theme.TextPrimary
import com.kasatria.kasatriaunknownbank.ui.theme.White

import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.rememberScrollState

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue


@Composable
fun CreditCardFilterScreen(
    filters: CreditCardFilters,
    onFiltersChange: (CreditCardFilters) -> Unit,
    onBack: () -> Unit,
    onShowResults: () -> Unit
) {

    var draftFilters by remember(filters) {
        mutableStateOf(filters)
    }

    val draftResultCount =
        CreditCardProducts.all
            .applyFilters(draftFilters)
            .size

    val hasDraftFilters =
        draftFilters != CreditCardFilters()

    val hasFilterChanges =
        draftFilters != filters

    Box(
        modifier = Modifier
            .fillMaxSize()
            .background(Color.White)
    ) {

        Column(
            modifier = Modifier
                .fillMaxSize()
                .statusBarsPadding()
        ) {

            /*
             * =================================
             * HEADER
             * =================================
             */

            Box(
                modifier = Modifier
                    .fillMaxWidth()
                    .height(56.dp)
            ) {

                IconButton(
                    onClick = onBack,
                    modifier = Modifier
                        .padding(start = 8.dp)
                        .align(Alignment.CenterStart)
                        .size(48.dp)
                ) {

                    Icon(
                        imageVector =
                            Icons.AutoMirrored.Filled.ArrowBack,

                        contentDescription = "Back",

                        tint = TextPrimary
                    )
                }


                Text(
                    text = "Filters",

                    modifier =
                        Modifier.align(
                            Alignment.Center
                        ),

                    color = TextPrimary,

                    fontSize = 13.sp,

                    fontWeight =
                        FontWeight.Bold
                )


                Text(
                    text = "Clear All",

                    modifier = Modifier
                        .align(Alignment.CenterEnd)
                        .padding(end = 20.dp)
                        .clickable(
                            enabled = hasDraftFilters
                        ) {
                            draftFilters =
                                CreditCardFilters()
                        },

                    color =
                        if (hasDraftFilters) {
                            PrimaryBlue
                        } else {
                            Color(0xFFAAAAAA)
                        },

                    fontSize = 13.sp,
                    fontWeight = FontWeight.Bold
                )
            }


            /*
             * =================================
             * SEARCH
             * =================================
             */

            FilterSectionHeader(
                title = "Search for"
            )


            Spacer(
                modifier =
                    Modifier.height(15.dp)
            )


            SearchField(
                value = draftFilters.searchText,
                onValueChange = {
                    draftFilters =
                        draftFilters.copy(
                            searchText = it
                        )
                },
                modifier = Modifier
                    .padding(horizontal = 20.dp)
                    .fillMaxWidth()
            )


            Spacer(
                modifier =
                    Modifier.height(30.dp)
            )


            FilterDivider()


            /*
             * =================================
             * CREDIT CARD TIERS
             * =================================
             */

            Spacer(
                modifier =
                    Modifier.height(20.dp)
            )


            FilterSectionHeader(
                title = "Credit Card Tiers"
            )


            Spacer(
                modifier =
                    Modifier.height(20.dp)
            )


            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .horizontalScroll(
                        rememberScrollState()
                    )
                    .padding(horizontal = 20.dp),

                horizontalArrangement =
                    Arrangement.spacedBy(10.dp)
            ) {

                listOf(
                    "All",
                    "Silver",
                    "Gold",
                    "World",
                    "Platinum"
                ).forEach { tier ->

                    FilterChip(
                        text = tier,
                        selected =
                            draftFilters.tier == tier,
                        onClick = {
                            draftFilters =
                                draftFilters.copy(
                                    tier = tier
                                )
                        }
                    )
                }
            }


            Spacer(
                modifier =
                    Modifier.height(30.dp)
            )


            FilterDivider()


            /*
             * =================================
             * INTEREST
             * =================================
             */

            Spacer(
                modifier =
                    Modifier.height(20.dp)
            )


            FilterSectionHeader(
                title = "Interest"
            )


            Spacer(
                modifier =
                    Modifier.height(20.dp)
            )


            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .horizontalScroll(
                        rememberScrollState()
                    )
                    .padding(horizontal = 20.dp),

                horizontalArrangement =
                    Arrangement.spacedBy(10.dp)
            ) {

                listOf(
                    "All",
                    "Islamic",
                    "Cashback",
                    "Travel",
                    "Rewards",
                    "Petrol",
                    "Dining"
                ).forEach { interest ->

                    FilterChip(
                        text = interest,
                        selected =
                            draftFilters.interest == interest,
                        onClick = {
                            draftFilters =
                                draftFilters.copy(
                                    interest = interest
                                )
                        }
                    )
                }
            }


            Spacer(
                modifier =
                    Modifier.height(10.dp)
            )


            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .horizontalScroll(
                        rememberScrollState()
                    )
                    .padding(horizontal = 20.dp),

                horizontalArrangement =
                    Arrangement.spacedBy(10.dp)
            ) {

                listOf(
                    "All",
                    "Silver",
                    "Gold",
                    "World",
                    "Platinum"
                ).forEach { tier ->

                    FilterChip(
                        text = tier,
                        selected =
                            draftFilters.tier == tier,
                        onClick = {
                            draftFilters =
                                draftFilters.copy(
                                    tier = tier
                                )
                        }
                    )
                }
            }
        }


        /*
         * =================================
         * BOTTOM BUTTON
         * =================================
         */

        Button(
            onClick = {
                onFiltersChange(draftFilters)
                onShowResults()
            },
            enabled = draftResultCount > 0 && hasFilterChanges,
            modifier = Modifier
                .align(Alignment.BottomCenter)
                .navigationBarsPadding()
                .padding(
                    start = 16.dp,
                    end = 16.dp,
                    bottom = 16.dp
                )
                .fillMaxWidth()
                .height(46.dp),

            shape =
                RoundedCornerShape(100.dp),

            colors =
                ButtonDefaults.buttonColors(
                    containerColor = PrimaryBlue,
                    contentColor = White
                )
        ) {

            Text(
                text =
                    if (draftResultCount == 1) {
                        "Show 1 Credit Card"
                    } else {
                        "Show $draftResultCount Credit Cards"
                    }
            )
        }
    }
}


@Composable
private fun FilterSectionHeader(
    title: String
) {

    Row(
        modifier = Modifier
            .fillMaxWidth()
            .padding(horizontal = 20.dp),

        horizontalArrangement =
            Arrangement.SpaceBetween,

        verticalAlignment =
            Alignment.CenterVertically
    ) {

        Text(
            text = title,

            color = Color(0xFF333333),

            fontSize = 15.sp,

            fontWeight =
                FontWeight.Bold
        )


        Icon(
            imageVector =
                Icons.Default.KeyboardArrowUp,

            contentDescription = null,

            tint = Color(0xFF666666),

            modifier =
                Modifier.size(20.dp)
        )
    }
}


@Composable
private fun SearchField(
    value: String,
    onValueChange: (String) -> Unit,
    modifier: Modifier = Modifier
) {

    val shape =
        RoundedCornerShape(8.dp)

    Row(
        modifier = modifier
            .height(46.dp)
            .background(
                color = White,
                shape = shape
            )
            .border(
                width = 1.dp,
                color = Color(0xFFDDDDDD),
                shape = shape
            )
            .padding(
                horizontal = 19.dp
            ),

        verticalAlignment =
            Alignment.CenterVertically
    ) {

        BasicTextField(
            value = value,

            onValueChange =
                onValueChange,

            modifier =
                Modifier.weight(1f),

            singleLine = true,

            textStyle =
                androidx.compose.ui.text.TextStyle(
                    color = TextPrimary,
                    fontSize = 15.sp
                ),

            decorationBox = {
                    innerTextField ->

                Box {

                    if (value.isEmpty()) {

                        Text(
                            text = "Credit Card",
                            color = Color(0xFF999999),
                            fontSize = 15.sp
                        )
                    }

                    innerTextField()
                }
            }
        )


        Icon(
            imageVector =
                Icons.Default.Search,

            contentDescription =
                "Search",

            tint =
                Color(0xFF999999),

            modifier =
                Modifier.size(24.dp)
        )
    }
}


@Composable
private fun FilterChip(
    text: String,
    selected: Boolean,
    onClick: () -> Unit
) {

    val shape =
        RoundedCornerShape(8.dp)

    Box(
        modifier = Modifier
            .height(36.dp)
            .then(
                if (selected) {

                    Modifier.background(
                        color = LinkBlue,
                        shape = shape
                    )

                } else {

                    Modifier.border(
                        width = 1.dp,
                        color = Color(0xFF666666),
                        shape = shape
                    )
                }
            )
            .clickable(
                onClick = onClick
            )
            .padding(
                horizontal = 15.dp
            ),

        contentAlignment =
            Alignment.Center
    ) {

        Text(
            text = text,

            color =
                if (selected) {
                    White
                } else {
                    Color(0xFF666666)
                },

            fontSize = 15.sp
        )
    }
}


@Composable
private fun FilterDivider() {

    Box(
        modifier = Modifier
            .fillMaxWidth()
            .height(1.dp)
            .background(
                Color(0xFFD9D9D9)
                    .copy(alpha = 0.30f)
            )
    )
}

File: unknownbank_flutter/lib/screens/credit_card_filter_screen.dart
import 'package:flutter/material.dart';

import '../models/credit_card_filters.dart';
import '../models/credit_card_product.dart';
import '../theme/app_colors.dart';

class CreditCardFilterScreen extends StatefulWidget {
  final CreditCardFilters filters;

  const CreditCardFilterScreen({
    super.key,
    required this.filters,
  });

  @override
  State<CreditCardFilterScreen> createState() =>
      _CreditCardFilterScreenState();
}

class _CreditCardFilterScreenState
    extends State<CreditCardFilterScreen> {
  late CreditCardFilters draftFilters;

  @override
  void initState() {
    super.initState();
    draftFilters = widget.filters;
  }

  int get resultCount => draftFilters
      .apply(CreditCardProducts.all)
      .length;

  bool get hasChanges =>
      draftFilters != widget.filters;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.white,
      body: SafeArea(
        child: Column(
          children: [
            SizedBox(
              height: 56,
              child: Stack(
                alignment: Alignment.center,
                children: [
                  Align(
                    alignment: Alignment.centerLeft,
                    child: Padding(
                      padding:
                          const EdgeInsets.only(left: 8),
                      child: IconButton(
                        onPressed: () {
                          Navigator.pop(context);
                        },
                        icon: const Icon(
                          Icons.arrow_back,
                          color: AppColors.textPrimary,
                        ),
                      ),
                    ),
                  ),

                  const Text(
                    'Filters',
                    style: TextStyle(
                      fontSize: 13,
                      fontWeight: FontWeight.bold,
                      color: AppColors.textPrimary,
                    ),
                  ),

                  Align(
                    alignment: Alignment.centerRight,
                    child: Padding(
                      padding:
                          const EdgeInsets.only(right: 20),
                      child: GestureDetector(
                        onTap: draftFilters.isDefault
                            ? null
                            : () {
                                setState(() {
                                  draftFilters =
                                      const CreditCardFilters();
                                });
                              },
                        child: Text(
                          'Clear All',
                          style: TextStyle(
                            fontSize: 13,
                            fontWeight: FontWeight.bold,
                            color: draftFilters.isDefault
                                ? const Color(0xFFAAAAAA)
                                : AppColors.primary,
                          ),
                        ),
                      ),
                    ),
                  ),
                ],
              ),
            ),

            Expanded(
              child: SingleChildScrollView(
                padding:
                    const EdgeInsets.only(bottom: 30),
                child: Column(
                  crossAxisAlignment:
                      CrossAxisAlignment.start,
                  children: [
                    const _SectionHeader(
                      title: 'Search for',
                    ),

                    const SizedBox(height: 15),

                    Padding(
                      padding:
                          const EdgeInsets.symmetric(
                        horizontal: 20,
                      ),
                      child: TextField(
                        onChanged: (value) {
                          setState(() {
                            draftFilters =
                                draftFilters.copyWith(
                              searchText: value,
                            );
                          });
                        },
                        controller: TextEditingController(
                          text:
                              draftFilters.searchText,
                        )
                          ..selection =
                              TextSelection.collapsed(
                            offset: draftFilters
                                .searchText.length,
                          ),
                        decoration:
                            const InputDecoration(
                          hintText: 'Credit Card',
                          suffixIcon:
                              Icon(Icons.search),
                          border: OutlineInputBorder(),
                          contentPadding:
                              EdgeInsets.symmetric(
                            horizontal: 19,
                          ),
                        ),
                      ),
                    ),

                    const SizedBox(height: 30),
                    const Divider(height: 1),
                    const SizedBox(height: 20),

                    const _SectionHeader(
                      title: 'Credit Card Tiers',
                    ),

                    const SizedBox(height: 20),

                    _FilterRow(
                      values: const [
                        'All',
                        'Silver',
                        'Gold',
                        'World',
                        'Platinum',
                      ],
                      selected: draftFilters.tier,
                      onSelected: (value) {
                        setState(() {
                          draftFilters =
                              draftFilters.copyWith(
                            tier: value,
                          );
                        });
                      },
                    ),

                    const SizedBox(height: 30),
                    const Divider(height: 1),
                    const SizedBox(height: 20),

                    const _SectionHeader(
                      title: 'Interest',
                    ),

                    const SizedBox(height: 20),

                    _FilterRow(
                      values: const [
                        'All',
                        'Islamic',
                        'Cashback',
                        'Travel',
                        'Rewards',
                        'Petrol',
                        'Dining',
                      ],
                      selected:
                          draftFilters.interest,
                      onSelected: (value) {
                        setState(() {
                          draftFilters =
                              draftFilters.copyWith(
                            interest: value,
                          );
                        });
                      },
                    ),
                  ],
                ),
              ),
            ),

            Padding(
              padding: const EdgeInsets.fromLTRB(
                16,
                8,
                16,
                16,
              ),
              child: SizedBox(
                width: double.infinity,
                height: 46,
                child: FilledButton(
                  onPressed:
                      resultCount > 0 && hasChanges
                          ? () {
                              Navigator.pop(
                                context,
                                draftFilters,
                              );
                            }
                          : null,
                  style: FilledButton.styleFrom(
                    backgroundColor:
                        AppColors.primary,
                  ),
                  child: Text(
                    resultCount == 1
                        ? 'Show 1 Credit Card'
                        : 'Show $resultCount Credit Cards',
                  ),
                ),
              ),
            ),
          ],
        ),
      ),
    );
  }
}

class _SectionHeader extends StatelessWidget {
  final String title;

  const _SectionHeader({
    required this.title,
  });

  @override
  Widget build(BuildContext context) {
    return Padding(
      padding:
          const EdgeInsets.symmetric(horizontal: 20),
      child: Row(
        mainAxisAlignment:
            MainAxisAlignment.spaceBetween,
        children: [
          Text(
            title,
            style: const TextStyle(
              fontSize: 15,
              fontWeight: FontWeight.bold,
              color: Color(0xFF333333),
            ),
          ),
          const Icon(
            Icons.keyboard_arrow_up,
            size: 20,
            color: Color(0xFF666666),
          ),
        ],
      ),
    );
  }
}

class _FilterRow extends StatelessWidget {
  final List<String> values;
  final String selected;
  final ValueChanged<String> onSelected;

  const _FilterRow({
    required this.values,
    required this.selected,
    required this.onSelected,
  });

  @override
  Widget build(BuildContext context) {
    return SingleChildScrollView(
      scrollDirection: Axis.horizontal,
      padding:
          const EdgeInsets.symmetric(horizontal: 20),
      child: Row(
        children: [
          for (int i = 0;
              i < values.length;
              i++) ...[
            _FilterChip(
              text: values[i],
              selected: selected == values[i],
              onTap: () {
                onSelected(values[i]);
              },
            ),
            if (i != values.length - 1)
              const SizedBox(width: 10),
          ],
        ],
      ),
    );
  }
}

class _FilterChip extends StatelessWidget {
  final String text;
  final bool selected;
  final VoidCallback onTap;

  const _FilterChip({
    required this.text,
    required this.selected,
    required this.onTap,
  });

  @override
  Widget build(BuildContext context) {
    return InkWell(
      onTap: onTap,
      borderRadius: BorderRadius.circular(8),
      child: Container(
        height: 36,
        alignment: Alignment.center,
        padding:
            const EdgeInsets.symmetric(horizontal: 15),
        decoration: BoxDecoration(
          color: selected
              ? AppColors.link
              : Colors.transparent,
          borderRadius: BorderRadius.circular(8),
          border: Border.all(
            color: selected
                ? AppColors.link
                : const Color(0xFF666666),
          ),
        ),
        child: Text(
          text,
          style: TextStyle(
            fontSize: 15,
            color: selected
                ? Colors.white
                : const Color(0xFF666666),
          ),
        ),
      ),
    );
  }
}